Sources/App/Commands/BuildTrigger.swift

```diff
@@ -0,0 +1,162 @@
+    fetchBuildCandidates(application.db, limit: limit)
+        .flatMap { triggerBuilds(on: application.db, client: application.client, packages: $0) }
+}
+
+
+func triggerBuilds(on database: Database,
+                   client: Client,
+                   packages: [Package.Id]) -> EventLoopFuture<Void> {
+    packages.map {
+        findMissingBuilds(database, packageId: $0)
+            .flatMap { triggers in
+                triggers.map { trigger -> EventLoopFuture<Void> in
+                    trigger.pairs.map { pair in
+                        Build.trigger(database: database,
+                                      client: client,
+                                      platform: pair.platform,
+                                      swiftVersion: pair.swiftVersion,
+                                      versionId: trigger.versionId)
+                            .flatMap { _ in
+                                Build(versionId: trigger.versionId,
+                                      platform: pair.platform,
+                                      status: .pending,
+                                      swiftVersion: pair.swiftVersion)
+                                    .create(on: database)
+                            }
+                            .transform(to: ())
+                    }
+                    .flatten(on: database.eventLoop)
+                }
+                .flatten(on: database.eventLoop)
+            }
+    }
+    .flatten(on: database.eventLoop)
+}
+
+
+func fetchBuildCandidates(_ database: Database,
+                          limit: Int) -> EventLoopFuture<[Package.Id]> {
+    guard let db = database as? SQLDatabase else {
+        fatalError("Database must be an SQLDatabase ('as? SQLDatabase' must succeed)")
+    }
+
+    struct Row: Decodable {
+        var packageId: Package.Id
+
+        enum CodingKeys: String, CodingKey {
+            case packageId = "package_id"
+        }
+    }
+
+    let versions = SQLIdentifier("versions")
+    let packageId = SQLIdentifier("package_id")
+    let latest = SQLIdentifier("latest")
+    let updatedAt = SQLIdentifier("updated_at")
+
+    let swiftVersions = SwiftVersion.allActive
+    let platforms = Build.Platform.allActive
+    let expectedBuildCount = swiftVersions.count * platforms.count
+
+    let buildUnderCount = SQLBinaryExpression(left: count(SQLRaw("*")),
+                                              op: SQLBinaryOperator.lessThan,
+                                              right: SQLLiteral.numeric("\(expectedBuildCount)"))
+
+    let query = db
+        .select()
+        .column(packageId)
+        .from(versions)
+        .join("builds", method: .left, on: "builds.version_id=versions.id")
+        .where(isNotNull(latest))
+        .groupBy(packageId)
+        .groupBy(latest)
+        .groupBy(SQLColumn(updatedAt, table: versions))
+        .having(buildUnderCount)
+        .orderBy(SQLColumn(updatedAt, table: versions))
+        .limit(limit)
```

Comment:
I’m tempted to replace this with the corresponding raw SQL query. I’m not sure the type safety buys us much here and it’s not a dynamic query like search where we profit from automatic quoting.

The raw query has the advantage that it’s easy to copy and run it for manual validation.